### ISSUE TITLE
Kubernetes-master charm is not replacing hostname with IP address when configuring the kubeapi advertise-address

### ISSUE THREAD

**Issue by CalvinHartwell** [juju-solutions/bundle-canonical-kubernetes#573]
Hi all, 

Though I opened an issue yesterday and closed it [https://github.com/juju-solutions/bundle-canonical-kubernetes/issues/572](https://github.com/juju-solutions/bundle-canonical-kubernetes/issues/572), our problem on VMware was not caused by CPU/memory constraints, but the way juju and the kubernetes-master charm are handling the hostname. 

On a working cluster, the service for kubernetes-api-server on a working kubernertes master looks like this:

```
ubuntu@ip-172-31-16-189:~$ ps -aux | grep kube-apiserver
root     25489  2.9  3.0 584568 483468 ?       Ssl  07:46   2:20 /snap/kube-apiserver/403/kube-apiserver --advertise-address 172.31.16.189 --allow-privileged=false --authorization-mode AlwaysAllow --basic-auth-file /root/cdk/basic_auth.csv --client-ca-file /root/cdk/ca.crt --enable-aggregator-routing --etcd-cafile /root/cdk/etcd/client-ca.pem --etcd-certfile /root/cdk/etcd/client-cert.pem --etcd-keyfile /root/cdk/etcd/client-key.pem --etcd-servers https://172.31.10.240:2379,https://172.31.28.68:2379,https://172.31.38.201:2379 --kubelet-certificate-authority /root/cdk/ca.crt --kubelet-client-certificate /root/cdk/client.crt --kubelet-client-key /root/cdk/client.key --kubelet-preferred-address-types [InternalIP,Hostname,InternalDNS,ExternalDNS,ExternalIP] --logtostderr --min-request-timeout 300 --proxy-client-cert-file /root/cdk/client.crt --proxy-client-key-file /root/cdk/client.key --requestheader-allowed-names client --requestheader-client-ca-file /root/cdk/ca.crt --requestheader-extra-headers-prefix X-Remote-Extra- --requestheader-group-headers X-Remote-Group
```

Note the --advertise-address of 172.31.16.189, this is correct for starting this service. However, when using the juju manual provider with SSH keys and hostnames, we get the following error: 

```
 unic@infkc-prd-mstr1:/var/log/juju$ journalctl | grep kube-api | grep Error
May 15 10:27:00 infkc-prd-mstr1.p.ilovecats.net kube-apiserver.daemon[30691]: Error: invalid argument "infkc-prd-mstr1.p.ilovecats.net" for "--advertise-address" flag: failed to parse IP: "infkc-prd-mstr1.p.ilovecats.net"
May 15 10:27:00 infkc-prd-mstr1.p.ilovecats.net kube-apiserver.daemon[30852]: Error: invalid argument "infkc-prd-mstr1.p.ilovecats.net" for "--advertise-address" flag: failed to parse IP: "infkc-prd-mstr1.p.ilovecats.net"
May 15 10:27:01 infkc-prd-mstr1.p.ilovecats.net kube-apiserver.daemon[31349]: Error: invalid argument "infkc-prd-mstr1.p.ilovecats.net" for "--advertise-address" flag: failed to parse IP: "infkc-prd-mstr1.p.ilovecats.net"
May 15 10:27:01 infkc-prd-mstr1.p.ilovecats.net kube-apiserver.daemon[31632]: Error: invalid argument "infkc-prd-mstr1.p.ilovecats.net" for "--advertise-address" flag: failed to parse IP: "infkc-prd-mstr1.p.ilovecats.net
```
Note that the charm is not applying validation to the hostname: it should realize it is not an IP address and attempt to resolve the hostname to an IP address. 

This is 100% reproducible, just bootstrap a cluster using cdk-core on 2 nodes with manual provider and DNS entries instead of IP addresses, we can help set this up if you require. 

We can run the cdk-debug tool (https://github.com/juju-solutions/cdk-field-agent) if required, let us know. We will replace the master nodes with IP address and try to redeploy.

**Comment by lolwww** [comment]
kubernetes/cmd/kubeadm/app/util/config/masterconfig.go

	// validate cfg.API.AdvertiseAddress.
	addressIP := net.ParseIP(cfg.API.AdvertiseAddress)
	if addressIP == nil && cfg.API.AdvertiseAddress != "" {
		return fmt.Errorf("couldn't use \"%s\" as \"apiserver-advertise-address\", must be ipv4 or ipv6 address", cfg.API.AdvertiseAddress)
	}

**Comment by tvansteenburgh** [comment]
Working on a fix for this.

**Comment by Cynerva** [comment]
Thanks @CalvinHartwell.

I believe this bug originates from Juju. Specifically here: https://bugs.launchpad.net/charm-nova-compute/+bug/1721368

Which has a fix merged 6 days ago: https://github.com/juju/juju/pull/8672

I don't think we should fix this in the charm. It's specific to the manual provider, has a simple (but time-consuming) workaround, and should be fixed in a Juju release soon anyway.

We do, however, need to verify that the above Juju PR does fix the problem for us.

**Comment by CalvinHartwell** [comment]
@tvansteenburgh @Cynerva customer says this was working a couple of weeks back so it does look fresh, thanks for quick replies guys.

**Comment by Cynerva** [comment]
Yeah that makes sense. We didn't start setting --advertise-address until very recently: https://github.com/kubernetes/kubernetes/pull/62169

Looks like that went out with kubernetes-master revision 102 on Apr 9.

**Comment by dshcherb** [comment]
@tvansteenburgh @Cynerva 

Same issue here.

When I added machines to the model via hostnames I started getting the same type of errors about advertise-address.

**Comment by carmine** [comment]
Latest comment from Dmitrii:

https://bugs.launchpad.net/juju/+bug/1721368/comments/20

**Comment by Cynerva** [comment]
Another PR has recently been merged in juju to fix this: https://github.com/juju/juju/pull/8885

Again, it's looking like this will be fixed in Juju 2.4.1 and 2.3.9, but I need to confirm that the linked PR actually fixes this.

**Comment by Cynerva** [comment]
I've confirmed that the issue is fixed on a dev build of Juju (2.4.1+2.4-be71bd3).
